node/core/approval-voting/src/lib.rs

```diff
@@ -1201,6 +1240,10 @@ async fn handle_from_overseer<Context>(
 
 			wakeups.prune_finalized_wakeups(block_number);
 
+			// `prune_finalized_wakeups` prunes all finalized block hashes. We prune spans accordingly.
+			let hash_set = wakeups.block_numbers.values().flatten().collect::<HashSet<_>>();
+			state.spans.retain(|hash, _| hash_set.contains(hash));
```

Comment:
Might be a premature optimization, but this looks expensive: we're recreating a set of all unfinalized blocks on every finality notification. Instead, we could pass state.spans to `prune_finalized_wakeups` to reuse the prunsed hash set of block hashes directly.

Reply:
Sounds reasonable!

Reply:
@ordian could you check if the logic makes sense for the suggested approach please?  In essence I want to retain the span , whenever `pruned_blocks` does not contain the hash. Note that `spans: HashMap<Hash, PerLeafSpan>`.

Reply:
yes, looks good!